doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1559 @@
+; field                        = value
+ADMIN_MODE                     = "enable" / "disable" ; If the NAT feature is enabled or disabled globally
+NAT_TIMEOUT                    = 1*6DIGIT         ; Timeout in secs (Range: 300 sec - 432000 sec)
+NAT_TCP_TIMEOUT                = 1*3DIGIT         ; Timeout in secs (Range: 300 sec - 432000 sec)
+NAT_UDP_TIMEOUT                = 1*6DIGIT         ; Timeout in secs (Range: 120 sec - 600 sec)
+```
+
+### 3.2.5 COUNTERS DB
+The following new counters are applicable per zone based on the support in the hardware.
+```
+COUNTERS_NAT_ZONE:<zone-id>
+    SAI_NAT_DNAT_DISCARDS
+    SAI_NAT_SNAT_DISCARDS
+    SAI_NAT_DNAT_TRANSLATION_NEEDED
+    SAI_NAT_SNAT_TRANSLATION_NEEDED
+    SAI_NAT_DNAT_TRANSLATIONS
+    SAI_NAT_SNAT_TRANSLATIONS
+
+```
+The following new counters are available per NAT/NAPT entry:
+```
+The counters in the COUNTERS_DB are updated every 5 seconds. The key for the entry in the COUNTERS_DB is same as the key in the APP_DB.
+
+COUNTERS_NAT:ip
+    DNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    DNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+    SNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    SNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+
+COUNTERS_NAPT:ip_protocol:ip:l4_port
+    DNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    DNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+    SNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    SNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+
+COUNTERS_NAT_TWICE:src_ip:dst_ip
+    DNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    DNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+    SNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    SNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+
+COUNTERS_NAPT_TWICE:ip_protocol:src_ip:src_l4_port:dst_ip:dst_l4_port
+    DNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    DNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+    SNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    SNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+```
+
+## 3.3 Switch State Service Design
+Following changes are done in the orchagent.
+
+### 3.3.1 NatMgr daemon
+ NatMgrd gets the STATIC_NAPT, STATIC_NAT, NAT_POOL, NAT_GLOBAL, NAT_BINDINGS config changes from CONFIG_DB.
+ NatMgr is responsible for pushing the Static NAT/NAPT entries and the NAT_GLOBAL configuration into the APP_DB. It also programs the Static NAT/NAPT entries and the NAT_POOL to ACL binding configuration as iptable rules in the kernel.
+
+Before acting upon the Static NAPT configuration, NatMgrd checks with the STATE_DB that the matching global IP interface is configured in the system (state == ok). 
+
+For a STATIC_NAPT entry and the interface configuration as below: 
+```
+STATIC_NAPT|65.55.42.1|TCP|1024 
+    "local_ip": 20.0.0.1
+    "local_port" :6000 
+    "nat_type": "dnat"
+
+INTERFACE|Ethernet15|65.55.42.1/24
+...
+
+INTERFACE|Ethernet15
+    "nat_zone": 1
```

Comment:
From this example, it seems configuring "nat_zone" at an interface is the way to enable this static DNAT on the interface., is it correct?  If there is another ethernet interface configured with a different nat_zone id, it'll not use DNAT entry, is it correct?  What if we want one more DNAT entry to be bound to ethernet15, how to do so?  Do we support two different static DNAT entries on two different interfaces with different zone-ids?  It is not clear to me how we can clearly associate which static NAT entries to which interface.

Reply:
The nat_zone mentioned here is used to push the zone attribute to the hardware on the L3 RIF (and only hardware uses it for the zone-id differences when doing the translation).

For binding the interface to the static dnat entry, the L3 IP (65.55.42.1) of the nat entry is matched against any L3 interface IP in the system. [This matching is needed for the application to add an entry to the hardware].
We can add any number of static napt entries with the same L3 IP (and with different ports) that will be matched to the same interface that has that IP (Ethernet15 here).
For example, in addition to the above static dnapt entry, we can add another static dnapt entry bound to the same interface.
STATIC_NAPT|65.55.42.1|TCP|1025 
    "local_ip": 30.0.0.1
    "local_port" :8000 
    "nat_type": "dnat"

Yes, we can add different static dnat entries matching different interfaces with same or different zone-ids.
For example, we have another L3 interface (Ethernet16) with IP 65.55.43.1, we can add another static dnapt entry
STATIC_NAPT|65.55.43.1|TCP|1024 
    "local_ip": 40.0.0.1
    "local_port" :6000 
    "nat_type": "dnat"

where 
INTERFACE|Ethernet16|65.55.43.1/24
...

and zone add can be 
INTERFACE|Ethernet16
    "nat_zone": 1

or
INTERFACE|Ethernet16
    "nat_zone": 2

Reply:
Did you mean the specific DNAT matching and action happens only if the original DIP matches the IP interface of the packets' incoming interface?  In example above, the DNAT action to translate from public IP 65.55.42.1|TCP|1024, will be applied to Ethernet15 only and not any other Ethernet interface, is it correct?  Or it is not interface specific?  E.g. if the same packet is received at another interface Ethernet17 with nat_zone 1 configured, but Ethernet17 has a different IP address configured, what will be the behavior?

Reply:
And what's the behavior when the configured DNAT global IP is matched to a loopback interface IP only?  thanks.

Reply:
> Did you mean the specific DNAT matching and action happens only if the original DIP matches the IP interface of the packets' incoming interface? In example above, the DNAT action to translate from public IP 65.55.42.1|TCP|1024, will be applied to Ethernet15 only and not any other Ethernet interface, is it correct? Or it is not interface specific? E.g. if the same packet is received at another interface Ethernet17 with nat_zone 1 configured, but Ethernet17 has a different IP address configured, what will be the behavior?

@rlhui Actually, matching against any interface in the same subnet is a criteria for application's purpose only to add the matching iptables rules in the kernel, and to push the entry to the hardware.
Once the entry is added in the hardware the behavior as per SAI pipeline, will not match on the packet's incoming interface. DNAT will happen for the packet ingressing with that original IP when received on any interface as long as DNAT criteria is met.

Reply:
@kirenkella, 
I see the following can happen with DNAT:
If the outside/public IP address to use is a local interface IP address (e.g. an uplink interface IP), in SONiC NAT spec, inbound traffic must be arriving at this interface, only then corresponding rule is added to iptables in kernel and hardware DNAT entry programmed.  Otherwise, it’ll not do so.
But in SAI pipeline, such check is not present, so long there is a “DNAT Pool Prefix Lookup” table hit which indicates the DIP needs a DNAT and if DNAT lookup is a miss, pkts will be trapped to CPU.
So if inbound traffic carrying this public IP is being received at a different uplink - an interface which does not own that public IP, no hardware DNAT entry is programmed but pkts will keep coming to CPU.

Will you please document this in the HLD so that those who use interface IP as public IP is aware of this and may avoid the situation.

Reply:
@rlhui, similar to loopback interfaces, the physical interface IP if used as public IP, can also be handled. Packet destined to interface IP received on any interface in the same zone as the interface that has that IP, is DNAT'ted if trapped to CPU. Section 3.4.1.1 is updated with the same.